src/rl78.c

```diff
@@ -40,48 +48,105 @@ static void rl78_set_reset(port_handle_t fd, int mode, int value)
     {
         serial_set_dtr(fd, level);
     }
+    }
 }
 
 int rl78_reset_init(port_handle_t fd, int wait, int baud, int mode, float voltage)
+// Routine for resetting RL78 and initializing it in programming mode
 {
     unsigned char r;
+    //(mode = desired communication mode from terminal input - 1)
+
+   // Determine UART mode from mode argument
     if (MODE_UART_1 == (mode & MODE_UART))
     {
         r = SET_MODE_1WIRE_UART;
         communication_mode = 1;
     }
+    
     else
     {
         r = SET_MODE_2WIRE_UART;
         communication_mode = 2;
     }
     if (4 <= verbose_level)
     {
+        if (mode == 4) // Special case for mode 5
+        {
+            communication_mode = 5;
+            printf("Using communication mode 5 (HW RESET with CTS reading RESET)\n");
+        }
+        else
+        {
         printf("Using communication mode %u%s\n",
                (mode & (MODE_UART | MODE_RESET)) + 1,
                (mode & MODE_INVERT_RESET) ? " with RESET inversion" : "");
+        }   
     }
-    rl78_set_reset(fd, mode, 0);                            /* RESET -> 0 */
-    serial_set_txd(fd, 0);                                  /* TOOL0 -> 0 */
-    if (wait)
+    // Begin reset procedure
+    if (mode == 4) // sequencing for mode 5
     {
-        printf("Turn MCU's power on and press any key...");
-        wait_kbhit();
-        printf("\n");
+        //printf("Turn MCU's power on with RESET low. Press any key to continue...\n");
+        //wait_kbhit();
+        serial_set_txd(fd, 0);
+        printf("Press and hold RESET.\n");
+        int reset = serial_get_cts(fd);
+        if (reset == 1)
+        {
+            while (1) // while CTS/RESET is high, wait
+            {
+                reset = serial_get_cts(fd);
+                if (reset == 0)
+                {
+                    printf("Got reset press. Hang on...\n");
+                    usleep(100000);
+                    printf("Let go of RESET.\n");
+                    break;
+                }
+            }
+        }
+        
+        while (1) // while CTS/RESET is low, wait
+        {
+            reset = serial_get_cts(fd);
+            if (reset == 1)
+            {
+               printf("Got reset release. Attempting to set communication mode...\n");
+               break;
+            }
+        }
```

Comment:
This is overly complex to read.

Replace the first block with just:
`while (serial_get_cts(fd)) { ; }`, then move the `printf(); usleep(); printf()` block outside the while loop.

Replace the second block with: `while (!serial_get_cts(fd)) { ; }` and move the `printf()` outside the while loop.